lib/Serialization/Deserialization.cpp

```diff
@@ -1120,11 +1120,13 @@ static void filterValues(Type expectedTy, ModuleDecl *expectedModule,
 
     if (isType != isa<TypeDecl>(value))
       return true;
-    auto ifaceTy = value->getInterfaceType();
-    if (!ifaceTy)
-      return true;
-    if (canTy && ifaceTy->getCanonicalType() != canTy)
-      return true;
+
+    // If we're expecting a type, make sure this decl has the expected type.
+    if (auto ifaceTy = value->getInterfaceType()) {
+      if (canTy && ifaceTy->getCanonicalType() != canTy)
+        return true;
+    }
```

Comment:
This doesn't make any sense either. You want it the other way around:

```cxx
if (canTy)
  if (auto ifaceTy = value->getInterfaceType())
    if (!ifaceTy->isEqual(canTy) )
      return true;
```

Reply:
Sorry, or just `if (canTy && !value->getInterfaceType()->isEqual(canTy))`. That was what I originally meant by "remove the filter".